pkg/kubectl/cmd/set/set_volume.go

```diff
@@ -0,0 +1,980 @@
+	client, err := f.ClientSet()
+	if err != nil {
+		return err
+	}
+	o.Client = client.Core()
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	mapper, _ := f.Object()
+
+	o.Output = cmdutil.GetFlagString(cmd, "output")
+	o.PrintObject = f.PrintObject
+
+	o.Cmd = cmd
+	o.DefaultNamespace = cmdNamespace
+	o.EnforceNamespace = enforceNamespace
+	o.Out = out
+	o.Err = errOut
+	o.Mapper = mapper
+	o.RESTClientFactory = f.ClientForMapping
+	o.UpdatePodSpecForObject = f.UpdatePodSpecForObject
+	o.Encoder = f.JSONEncoder()
+	o.DryRun = cmdutil.GetDryRunFlag(o.Cmd)
+
+	// In case of volume source ignore the default volume type
+	if len(o.AddOpts.Source) > 0 {
+		o.AddOpts.Type = ""
+	}
+	if len(o.AddOpts.ClaimSize) > 0 {
+		o.AddOpts.CreateClaim = true
+		if len(o.AddOpts.ClaimName) == 0 {
+			o.AddOpts.ClaimName = names.SimpleNameGenerator.GenerateName("pvc-")
+		}
+		q, err := kresource.ParseQuantity(o.AddOpts.ClaimSize)
+		if err != nil {
+			return fmt.Errorf("--claim-size is not valid: %v", err)
+		}
+		o.AddOpts.ClaimSize = q.String()
+	}
+	if len(o.AddOpts.DefaultMode) == 0 {
+		o.AddOpts.DefaultMode = "644"
+	}
+	switch strings.ToLower(o.AddOpts.ClaimMode) {
+	case strings.ToLower(string(api.ReadOnlyMany)), "rom":
+		o.AddOpts.ClaimMode = string(api.ReadOnlyMany)
+	case strings.ToLower(string(api.ReadWriteOnce)), "rwo":
+		o.AddOpts.ClaimMode = string(api.ReadWriteOnce)
+	case strings.ToLower(string(api.ReadWriteMany)), "rwm":
+		o.AddOpts.ClaimMode = string(api.ReadWriteMany)
+	case "":
+	default:
+		return errors.New("--claim-mode must be one of ReadWriteOnce (rwo), ReadWriteMany (rwm), or ReadOnlyMany (rom)")
+	}
+	return nil
+}
+
+func (o *VolumeOptions) RunVolume(args []string, f cmdutil.Factory) error {
+	builder := f.NewBuilder().
+		ContinueOnError().
+		NamespaceParam(o.DefaultNamespace).DefaultNamespace().
+		FilenameParam(o.EnforceNamespace, &o.FilenameOptions).
+		Flatten()
+
+	if !o.Local {
+		builder = builder.
+			SelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, args...).
+			Latest()
+	} else {
+		// if a --local flag was provided, and a resource was specified in the form
+		// <resource>/<name>, fail immediately as --local cannot query the api server
+		// for the specified resource.
+		if len(args) > 0 {
+			return resource.LocalResourceError
+		}
+
+		builder = builder.Local(f.ClientForMapping)
+	}
+
+	singleItemImplied := false
+	infos, err := builder.Do().IntoSingleItemImplied(&singleItemImplied).Infos()
+	if err != nil {
+		return err
+	}
+
+	if o.List {
+		listingErrors := o.printVolumes(infos)
+		if len(listingErrors) > 0 {
+			return cmdutil.ErrExit
+		}
+		return nil
+	}
+
+	updateInfos := []*resource.Info{}
+	// if a claim should be created, generate the info we'll add to the flow
+	if o.Add && o.AddOpts.CreateClaim {
+		claim := o.AddOpts.createClaim()
+		m, err := o.Mapper.RESTMapping(api.Kind("PersistentVolumeClaim"))
```

Comment:
@zjj2wry @liggitt I got trouble here, can't find Kind function in "k8s.io/api/core/v1"

Reply:
GroupVersionKind.GroupKind()

Reply:
@zjj2wry can you describe it in detail?thanks a lot

Reply:
`mapping, err := mapper.RESTMapping(schema.GroupKind{Kind: "PersistentVolumeClaim"})` maybe can work